pvlib/iotools/srml.py

```diff
@@ -0,0 +1,148 @@
+"""Collection of functions to get data from UO SRML.
+"""
+import numpy as np
+import pandas as pd
+
+
+def read_srml(filename):
+    """
+    Read SRML file into pandas dataframe.
+
+    Parameters
+    ----------
+    filename: str
+        filepath or url to read for the tsv file.
+
+    Returns
+    -------
+    data: Dataframe
+        A dataframe with datetime index and all of the variables listed
+        in the `var_map` dict inside of the map_columns function, along
+        with their associated quality control flags.
+
+    Notes
+    -----
+    Note that the time index is shifted back one minute to account for
+    2400 hours, and to avoid time parsing errors on leap years. Data values
+    on a given line should now be understood to occur during the interval
+    extending from the time of the line in which they are listed to
+    the ending time on the next line, rather than the previous line.
+    """
+    tsv_data = pd.read_csv(filename, delimiter='\t')
+    year = tsv_data.columns[1]
+    data = format_index(tsv_data, year)
+    # Rename and drop datetime columns
+    data = data[data.columns[2:]].rename(columns=map_columns)
+
+    # Quality flags are all labeled 0, but occur immediately after their
+    # associated var so we create a dict mapping them to var_flag for renaming
+    columns = data.columns
+    flag_label_map = {flag: columns[columns.get_loc(flag) - 1] + '_flag'
+                      for flag in columns[1::2]}
+    data = data.rename(columns=flag_label_map)
+    # For data flagged bad or missing, replace the value with np.NaN
+    for col in columns[::2]:
+        data[col] = data[col].where(~(data[col + '_flag'] == 99), np.NaN)
+    return data
+
+
+def map_columns(col):
+    """Map column labels to pvlib names.
+
+    Parameters
+    ----------
+    col: str
+        Column label to be mapped.
+
+    Returns
+    -------
+    str
+        The pvlib label if it was found in the mapping,
+        else the original label.
+
+    Notes
+    -----
+    var_map is a dictionary mapping SRML data element numbers
+    to their pvlib names. For most variables, only the first
+    three digits are used, the fourth indicating the instrument.
+    Spectral data (7xxx) uses all four digits to indicate the
+    variable.
+    """
+    var_map = {
```

Comment:
this is ok but I am wondering if it would be better to define it at the top of the module. Then it's easier for people to inspect it if they are interested.

Reply:
I agree, I must have misunderstood something the other day, I'll replace this at the top of the module.